packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Providers/ModelProvider.cs

```diff
@@ -235,14 +235,46 @@ private ConstructorProvider BuildFullConstructor()
 
         private ValueExpression GetExpression(ParameterProvider parameter)
         {
-            if (parameter.Property is not null && parameter.Property.IsDiscriminator)
+            if (parameter.Property is not null && parameter.Property.IsDiscriminator && _inputModel.DiscriminatorValue != null)
             {
-                return IsUnknownDiscriminatorModel ? NullCoalescing(parameter.AsExpression, Literal(_inputModel.DiscriminatorValue)) : Literal(_inputModel.DiscriminatorValue);
+                var type = parameter.Property.Type;
+                if (IsUnknownDiscriminatorModel)
+                {
+                    if (!type.IsFrameworkType && type.IsEnum)
+                    {
+                        if (type.IsStruct)
+                        {
+                            /* kind != default ? kind : "unknown" */
+                            return new TernaryConditionalExpression(parameter.AsExpression.NotEqual(Default), parameter.AsExpression, Literal(_inputModel.DiscriminatorValue));
+                        }
+                        else
+                        {
+                            return parameter.AsExpression;
+                        }
+                    }
+                    else
+                    {
+                        /* kind ?? "unknown" */
+                        return NullCoalescing(parameter.AsExpression, Literal(_inputModel.DiscriminatorValue));
+                    }
+                }
+                else
+                {
+                    if (!type.IsFrameworkType && type.IsEnum)
+                    {
+                        var enumMember = type.EnumTypeMembers.FirstOrDefault(e => e.Value.ToString() == _inputModel.DiscriminatorValue) ?? throw new InvalidProgramException($"invalid discriminator value {_inputModel.DiscriminatorValue}");
```

Comment:
Should this be InvalidOperationException?

Reply:
Thanks @JoshLove-msft . updated.